File: apps/ios/Pebbles/Features/Profile/Models/Collection.swift
import Foundation

/// Mode variants for a collection. Mirrors the `mode` check constraint on
/// `public.collections`: `('stack', 'pack', 'track')` or `null`.
enum CollectionMode: String, Decodable, CaseIterable, Hashable {
    case stack
    case pack
    case track
}

/// View-model for a collection row. Not the storage shape — `pebbleCount` comes
/// from a PostgREST nested aggregate (`pebble_count:collection_pebbles(count)`)
/// that returns `[{ "count": N }]`. The custom decoder below unwraps that into
/// a plain `Int`, and falls back to `0` when the aggregate is absent (e.g. a
/// single-row fetch for the detail header).
struct Collection: Identifiable, Decodable, Hashable {
    let id: UUID
    let name: String
    let mode: CollectionMode?
    let pebbleCount: Int

    enum CodingKeys: String, CodingKey {
        case id, name, mode
        case pebbleCount = "pebble_count"
    }

    private struct CountWrapper: Decodable { let count: Int }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try container.decode(UUID.self, forKey: .id)
        self.name = try container.decode(String.self, forKey: .name)
        self.mode = try container.decodeIfPresent(CollectionMode.self, forKey: .mode)

        if let wrappers = try? container.decode([CountWrapper].self, forKey: .pebbleCount) {
            self.pebbleCount = wrappers.first?.count ?? 0
        } else {
            self.pebbleCount = 0
        }
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/profile/models/Collection.kt
package app.pbbls.android.features.profile.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Mode variants for a collection — mirrors the `mode` check constraint on
 * `public.collections`: `('stack', 'pack', 'track')` or null.
 */
@Serializable
enum class CollectionMode {
    @SerialName("stack")
    STACK,

    @SerialName("pack")
    PACK,

    @SerialName("track")
    TRACK,
}

/**
 * Clean domain type for a collection with its live pebble count — ports iOS
 * `Collection.swift` (D10). Distinct from the form's bare
 * [app.pbbls.android.features.path.models.PebbleCollection] `{id, name}`
 * reference model; list/detail/profile surfaces use this richer shape.
 */
data class Collection(
    val id: String,
    val name: String,
    val mode: CollectionMode?,
    val pebbleCount: Int,
)

/**
 * Wire row for `collections` selects with the
 * `pebble_count:collection_pebbles(count)` nested aggregate, which PostgREST
 * returns as `[{ "count": N }]`; an absent aggregate (e.g. a single-row detail
 * fetch) decodes to 0 defensively. Note the junction is `collection_pebbles`
 * — the opposite word order from `pebble_souls`.
 */
@Serializable
data class CollectionRow(
    val id: String,
    val name: String,
    val mode: CollectionMode? = null,
    @SerialName("pebble_count")
    val pebbleCount: List<CountRow> = emptyList(),
) {
    fun toCollection(): Collection =
        Collection(
            id = id,
            name = name,
            mode = mode,
            pebbleCount = pebbleCount.firstOrNull()?.count ?: 0,
        )

    @Serializable
    data class CountRow(
        val count: Int,
    )
}
